### ISSUE TITLE
Plotting connectivity matrices with both negative and positive values should change default colorbar

### ISSUE THREAD

**Issue by JohannesWiesner** [mne-tools/mne-connectivity#248]
If your issue is a usage question, please consider asking on the
[MNE Forum](https://mne.discourse.group) instead of opening an issue.

#### Describe the problem

The default colormap for mne-connectivity.viz.plot_connectivity_circle is "hot". This colorscale only makes sense for positive values. However, my matrix contains both positive and negative values. In this case the colormap should change to a diverging map with two different colors mapping to positive and negative values (e.g. blue to red with black in the middle). 0s should correspond to the background color.

#### Describe your solution

I currently do this:

```python
    # If the matrix contains both negative and positive values then 
    # the colorscale has to have two different colors with 0 corresponding
    # to the background color
    facecolor = 'white'
    textcolor = 'black'
    
    if np.any(matrix_values > 0) & np.any(matrix_values < 0):
    
        # Find the minimum and maximum values in the data and compute
        # midpoint between vmin and vmax where 0 should lie
        vmin = matrix_values.min()
        vmax = matrix_values.max()
        midpoint = abs(vmin) / (vmax - vmin)
    
        colors = [
            (0, "blue"),  # Negative values: blue
            (midpoint, facecolor),  # Zero: black
            (1, "red")  # Positive values: red
        ]
    
        # Create the colormap
        cmap = LinearSegmentedColormap.from_list("custom_cmap",colors)
    
    else:
        cmap = 'hot'
```

**Comment by tsbinns** [comment]
There is precedent with things like [`plot_evoked_topomap`](https://mne.tools/stable/generated/mne.viz.plot_evoked_topomap.html) in the main package for having the default colourmap adjust to the values:

> ```If None, 'Reds' is used for data that is either all-positive or all-negative, and 'RdBu_r' is used otherwise.```

How do people feel about adding similar behaviour to `plot_connectivity_circle`?

But @JohannesWiesner, could the code you show not be replaced with just passing `"RdBu_r"` to the `colormap` parameter?

**Comment by larsoner** [comment]
> How do people feel about adding similar behaviour to plot_connectivity_circle?

Sounds good to me!

**Comment by tsbinns** [comment]
@JohannesWiesner Would you be able to open a PR for this?

**Comment by JohannesWiesner** [comment]
> But @JohannesWiesner, could the code you show not be replaced with just passing `"RdBu_r"` to the `colormap` parameter?

@tsbinns : This has two problems. 

1.)  I think it would be good if the value 0 always is the same as the "facecolor" argument, so that connections with the value 0 blend in with the background. Therefore the function has to know what the background color and then build a colorscale while taking the information into account

2.) If you're distribution is skewed, setting "RdBu_r" will not result in a colorscale that matches `0` to the background color.

See (this is what happens if you only set "RdBu_r"):

<img width="584" alt="image" src="https://github.com/user-attachments/assets/69db1b61-daf4-4e85-bab9-8328ec6d0825">

**Comment by JohannesWiesner** [comment]
My personal preference is that 0s are always represented with the background color so that as soon as I export the figure with a transparent color, these connections will not be shown. But of course that might be personal preference (for some users, the value 0 might contain information that should be visible). Notice however, that just choosing "RdBu_r" as default interferes with the facecolor argument. For example if the user sets the facecolor to white it gives you the illusion that some connections don't exist:

<img width="584" alt="image" src="https://github.com/user-attachments/assets/f2852208-91fd-480a-b024-759e0e50e39f">

when you set the facecolor to "black" suddenly these connections pop up:

<img width="584" alt="image" src="https://github.com/user-attachments/assets/0174dba7-e669-4669-a87a-09637091a4e0">

However the latter figure is not very useful in a manuscript because you want the node names to be printed in black

**Comment by tsbinns** [comment]
I definitely get that depending on facecolor the legibility of the colourmap changes, e.g., with the hot cmap those very strong values could become difficult to see with a white background:
![image](https://github.com/user-attachments/assets/18c21afb-3b3b-43f4-9ed6-730a192d46e1)

The question is whether there should be logic within the `plot_connectivity_circle` function to handle the interaction between facecolor and colourmap, or whether this is the responsibility of the user by using the available parameters.

I do not have the authority to decide on this though.

**Comment by tsbinns** [comment]
> Asked this question on Stackoverflow, maybe there is a generic way to customize a default colormap instead of building one from scratch

Cool, would be interested to see.

I came across [one convenient way](https://stackoverflow.com/a/62251907), but I don't know how generalisable this is across different ways of creating the colourbar.

In any case for the connectivity visualisation with divergent colourbars, some logic to center it around zero would be a nice feature.

**Comment by JohannesWiesner** [comment]
Maybe it would also make sense to open another issue? Wouldn't it make more sense to set the background color by default to white and the text color by default to black? Because this is closer to what users would also copy and paste into their text editors / poster editors? From there on, it would make more sense to think about  good default colormaps for positive-only and positive-and-negative connectivity matrices?

**Comment by JohannesWiesner** [comment]
> The question is whether there should be logic within the plot_connectivity_circle function to handle the interaction between facecolor and colourmap, or whether this is the responsibility of the user by using the available parameters.

Good point. In case of the latter, it might probably become necessary that the function can accept a `norm` argument and passes this one to the internal functions.

**Comment by drammock** [comment]
I agree with https://github.com/mne-tools/mne-connectivity/issues/248#issuecomment-2435902653 that auto-choosing `Reds` vs `RdBu_r` is fine, and we can/should leverage what we already have:

https://github.com/mne-tools/mne-python/blob/e15292fc0bc8d5e32dd6d6099a839bf810963f3a/mne/viz/utils.py#L1422-L1430

(which I guess means we should make that function public, if we're going to use it here).

I think it would be too strict to always force the center of a two-slope colormap to be the same as the facecolor.  Seems like a nice clear docs example is the best solution, maybe also a note in the Notes section of the docstring.  Here's a good starting point:

https://github.com/mne-tools/mne-python/blob/e15292fc0bc8d5e32dd6d6099a839bf810963f3a/examples/time_frequency/time_frequency_erds.py#L79-L90

**Comment by tsbinns** [comment]
> I think it would be too strict to always force the center of a two-slope colormap to be the same as the facecolor. Seems like a nice clear docs example is the best solution, maybe also a note in the Notes section of the docstring.

Okay, but based off of that example, this means there is also room to add a `cnorm` parameter to the function?

**Comment by drammock** [comment]
yeah sorry I meant to say that explicitly. Adding `cnorm` seems like the right move here.

**Comment by tsbinns** [comment]
So @JohannesWiesner, is this something you are able to create a PR for at this time? Or is there also some help we can offer?

**Comment by JohannesWiesner** [comment]
> So @JohannesWiesner, is this something you are able to create a PR for at this time? Or is there also some help we can offer?

I don't have time for this at the moment, but maybe over the Christmas holidays. 

Just a quick sanity check: if the matrix data is normally distributed, shouldn't a `TwoSlopeNorm` `RdBu_r` color scale automatically converge to a “normal” `RdBu_r` color scale? If so, and if there are [future plans](https://github.com/mne-tools/mne-connectivity/issues/248#issuecomment-2442160407) to make the automatic selection of color scales for matrices with one or two signs “smarter”, the latter could always be the TwoSlopeNorm of `RdBu_r`. 

On top, I would still pitch for a default white background with black text. In that case, `0` values for both the one-sign / two-sign colorscales should always correspond to the white background color.
